### PR TITLE
feat: datasets panel

### PR DESCRIPTION
**Pull request by mscolnick**
## 📝 Summary

This adds a new panel that intelligently discovers datasets (right now pandas, polars, pyarrow). and will show its shape and schema. From here you can explore its columns and get quick charts based on them - which you can add directly to your notebook. 

The goal is to extend this to external datasources and other types (DBs, etc)

## 🔍 Description of Changes

image

### CODE REVIEW

File: frontend/src/components/editor/chrome/wrapper/sidebar.tsx [hunk 1]
```diff
@@ -39,6 +39,13 @@ export const Sidebar: React.FC = () => {
       >
         {renderIcon("dependencies")}
       </SidebarItem>
+      <SidebarItem
+        tooltip="Data sources"
```
Comment (akshayka) [inline frontend/src/components/editor/chrome/wrapper/sidebar.tsx:43]:
```suggestion
        tooltip="Explore data sources"
```

File: frontend/src/components/editor/chrome/wrapper/sidebar.tsx [hunk 2]
```diff
@@ -39,6 +39,13 @@ export const Sidebar: React.FC = () => {
       >
         {renderIcon("dependencies")}
       </SidebarItem>
+      <SidebarItem
```
Comment (akshayka) [inline frontend/src/components/editor/chrome/wrapper/sidebar.tsx:42]: Can you move this panel up one? So it's right after "Explore Variables"

File: frontend/src/components/editor/chrome/panels/datasources-panel.tsx [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+import type { TopLevelFacetedUnitSpec } from "@/plugins/impl/data-explorer/queries/types";
+import { useTheme } from "@/theme/useTheme";
+import {
+  maybeAddAltairImport,
+  maybeAddMarimoImport,
+} from "@/core/cells/add-missing-import";
+import { autoInstantiateAtom } from "@/core/config/config";
+import { DataColumnPreview } from "@/core/kernel/messages";
+
+export const DataSourcesPanel: React.FC = () => {
```
Comment (akshayka) [inline frontend/src/components/editor/chrome/panels/datasources-panel.tsx:37]: Could it be possible to suggest or even prompt users to install `altair` if it's not installed, to get plots?

Doesn't have to be in this PR, but could be a good follow-up to make the plots feature more discoverable.

Comment (mscolnick) [reply frontend/src/components/editor/chrome/panels/datasources-panel.tsx:37]: easy to add with how we send back errors - i can do in a followup though

File: frontend/src/components/editor/chrome/panels/datasources-panel.tsx [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+        <span className="text-sm">{table.name}</span>
+      </div>
+      {renderRowsByColumns()}
+      <Button
+        className="group-hover:inline-flex hidden"
+        variant="text"
+        size="icon"
+        onClick={Events.stopPropagation(() => onAddTable(table))}
+      >
+        <PlusSquareIcon className="h-3 w-3" />
```
[the remark this answers is not in the document]
Comment (mscolnick) [reply frontend/src/components/editor/chrome/panels/datasources-panel.tsx]: done

**Review by akshayka — APPROVED** [review]

File: marimo/_data/charts.py [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+from typing import Any, cast
+
+from marimo._data.models import DataType
+
+
+@abc.abstractmethod
+class ChartBuilder:
+    @abc.abstractmethod
+    def altair(self, data: Any, column: str) -> Any:
+        raise NotImplementedError
```
[the remark this answers is not in the document]
Comment (mscolnick) [reply marimo/_data/charts.py:16]: ah very cool. i think for readability (and possibly they could diverge in the future), ill leave for now

**Pull request merged**